pipeline/run_models.py

```diff
@@ -0,0 +1,164 @@
+) -> dict[Model, Any]:
+    """
+    Creates every model from models.py, fits them, saves
+    them to pickle, saves best params and returns dict
+    mapping model names to the fitted model.
+
+    use joblib for saving models to file:
+    # https://scikit-learn.org/stable/model_persistence.html
+
+    Args:
+        x_train: Training dataset
+        y_train: Target for training
+
+    Returns: dictionary of model name to trained model model
+    """
+
+    # define a list of models and their corresponding grid search functions (from models.py)
+    model_jobs = [
+        (Model.MLP, create_mlp_grid_search),
+        (Model.RF, random_forest_regressor_gridsearch),
+        (Model.XGB, xgboost_classifier_gridsearch),
+        (Model.LOGREG, logistic_regression_gridsearch),
+        # (Model.STATMODEL, statistical_model), # TODO: Does not work currently...
+    ]
+
+    models: dict[Model, Any] = {}  # model name to the trained model
+
+    with open("models/training_results.txt", "a") as best_model_params_f:
+        # loop through each model and perform grid search
+        for model_name, model_func in model_jobs:
+            best_model, best_params = model_func(x_train, y_train)
+            best_model_params_f.write(  # save the best params to file
+                f"\nmodel: {model_name.value}, params: {best_params}"
+            )
+            joblib.dump(  # save the model as joblib file
+                best_model, f"{model_name.value}_best_model.joblib"
+            )
+            models[model_name] = best_model
+
+    return models
+
+
+def append_predictions_to_df(
+    df: pd.DataFrame, predictions: np.ndarray, model: Model
+) -> pd.DataFrame:
+    """
+    Save predictions made to dataframe.
+    @param df: Dataframe to append predictions to.
+    @param predictions: Predictions made by a model
+    @param model: Which model made the predictions
+    @return: Annoted dataframe
+    """
+
+    # The first test_size number of rows are used for testing
+    # ex. 1000 row dataframe might have a test_size of 200 rows
+    # to append the 200 rows of predictions, padding is needed
+    # so that it is the same length.
+    # so, pad to same length and append.
+    predictions_padded = np.pad(
+        predictions,
+        (0, len(df) - len(predictions)),
+        mode="constant", # type: ignore
+        constant_values=None,
+    )
+    col_name = f"{model.value}_preds"
+    df[col_name] = pd.Series(predictions_padded)
+    return df
+
+
+def test_models(
+    models: dict[Model, Any], x_test: np.ndarray, y_test: np.ndarray, df: pd.DataFrame
+) -> dict[str, dict]:
+    """
+    Tests all the models. Will return scoring metrics for each models predictions.
+
+    Args:
+        models (dict[Model, Any]): The dictionary of the best models after fitting on the train data.
+        x_test (np.ndarray): The input test data from the train-test split
+        y_test (np.ndarray): The target test data from the train-test split
+    """
```

Comment:
Missing return type in docstring:
```suggestion
       dict[str, dict]: Returns dictionary of scoring metrics for each model.
    """
```